### PR TITLE
Fix use of uninitialized memory for Vector3 constants

### PR DESCRIPTION
**Pull request by jkotas**
Found while investigating #72149. 

In a good case, it made the size and content of readonly data non-deterministic.
In a bad case, it exposed downstream bugs like #72149 intermittently.

### CODE REVIEW

**Comment by EgorBo** [comment]
I wonder if it should always be zero in the first place, it looks like that field is not zero-initialized with GenTreeVecCon is created https://github.com/dotnet/runtime/blob/51f3ca60ed119375fe55fd300b219ee2317dbb75/src/coreclr/jit/gentree.h#L3510-L3511

It looks like this is the place where it gets garbage value in the 4th element: https://github.com/dotnet/runtime/blob/51f3ca60ed119375fe55fd300b219ee2317dbb75/src/coreclr/jit/assertionprop.cpp#L3186-L3187 since we create a non-zero initialized veccon tree and set only 12 bytes

**Comment by jakobbotsch** [comment]
It is technically UB to access the wrong union member there, so I think the code should rather be changed to access `gtSimd12Val` and widen it explicitly. There is a similar occurrence in instr.cpp it seems.

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -2412,8 +2412,12 @@ void CodeGen::genSetRegToConst(regNumber targetReg, var_types targetType, GenTre
                         // Get a temp integer register to compute long address.
                         regNumber addrReg = tree->GetSingleTempReg();
 
-                        simd16_t             constValue = vecCon->gtSimd16Val;
-                        CORINFO_FIELD_HANDLE hnd        = emit->emitSimd16Const(constValue);
+                        simd16_t constValue = vecCon->gtSimd16Val;
+
+                        if (tree->TypeGet() == TYP_SIMD12)
+                            constValue.u32[3] = 0;
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegenarm64.cpp:2418]:
```suggestion
                        simd16_t constValue = {};

                        if (tree->TypeIs(TYP_SIMD12))
                            memcpy(&constValue, &tree->gtSimd12Val, sizeof(simd12_t));
                        else
                            constValue = tree->gtSimd16Val;
```

**Comment by jkotas** [comment]
> I wonder if it should always be zero in the first place, it looks like that field is not zero-initialized when GenTreeVecCon is created

I am not following. The line that you have highlighted is initializing `gtSimdBaseJitType` field. This PR is fixing uninitialized value of `gtSimd*Val` field.

**Comment by jakobbotsch** [comment]
This should get the same fix:
https://github.com/dotnet/runtime/blob/9e8d261ade2ee35cb679845af229d60c5c429304/src/coreclr/jit/instr.cpp#L765-L770

**Review by jakobbotsch — APPROVED** [review]

**Comment by EgorBo** [comment]
> I am not following. The line that you have highlighted is initializing gtSimdBaseJitType field. 

@jkotas That's the point - it doesn't initialize `gtSimd32Val` there leaving, potentially, garbage in it, does it?

**Comment by jkotas** [comment]
> it doesn't initialize gtSimd32Val there leaving, potentially, garbage in it, does it?

Right. The type of the tree is `TYP_SIMD12`, so one can naturally expect that the value is stored in gtSimd12Val field.

I think it would be unnatural to require that the top 4 bytes of gtSimd16Val field have to be initialized for `TYP_SIMD12` values.

**Pull request merged**